Fix macOS build by defaulting `VCPKG_TARGET_TRIPLET` to `x64-osx-min10.12`

fwcd opened the pull request:
Since the buildenv only provides dependencies for the `x64-osx-min10.12` triplet, building Mixxx on macOS using the `README.md` instructions (through `tools/macos_buildenv.sh setup`, with vcpkg using the `x64-osx` triplet by default) results in obscure missing library errors, like for example the following:

<details>

```
-- Could NOT find KeyFinder (missing: KeyFinder_LIBRARY KeyFinder_INCLUDE_DIR) (Required is at least version "2.2.6")
CMake Error at /Applications/CMake.app/Contents/share/cmake-3.23/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find FFTW (missing: FFTW_LIBRARY FFTW_INCLUDE_DIR)
Call Stack (most recent call first):
  /Applications/CMake.app/Contents/share/cmake-3.23/Modules/FindPackageHandleStandardArgs.cmake:594 (_FPHSA_FAILURE_MESSAGE)
  cmake/modules/FindFFTW.cmake:56 (find_package_handle_standard_args)
  buildenv/mixxx-deps-2.4-x64-osx-min10.12-3ede1a3/scripts/buildsystems/vcpkg.cmake:824 (_find_package)
  CMakeLists.txt:1988 (find_package)


-- Configuring incomplete, errors occurred!
See also "/Users/<USER>/git/mixxx/build/CMakeFiles/CMakeOutput.log".
See also "/Users/<USER>/git/mixxx/build/CMakeFiles/CMakeError.log".
```

</details>

While the issue can be worked around by manually setting `VCPKG_DEFAULT_TRIPLET` (like the CI workflows do too), it would be nice (for developers using macOS) if `cmake -B build` could "just work" after running the `tools/macos_buildenv.sh setup` script. This PR fixes the issue by defaulting to `x64-osx-min10.12` in the CMakeLists.

cc @daschuer

fwcd committed:
Set VCPKG_TARGET_TRIPLET on macOS by default

fwcd committed:
Only set VCPKG_TARGET_TRIPLET if unset

Swiftb0y reviewed (commented):
Looks like a decent solution to me, but I'm not familiar with the build process or mac os. What do you think @daschuer?

daschuer commented:
Thank you for this PR this alone looks as already good to me. However this reveals an inconsistency with CMAKE_OSX_DEPLOYMENT_TARGET

It is already set here:
https://github.com/mixxxdj/vcpkg/blob/3ede1a3c07a7a1b6fe979ba6ad3f9bffaa16283e/overlay/triplets/x64-osx-min10.12.cmake#L16

Which one is actually taken? 
Is the one in the CMakeList redundant?

daschuer commented:
Does this now fail early if the ripplet is not found in the vcpkg install directory? 
If not we may consider to add a sanity check. 

I think this also needs to be moved under the 
`if(DEFINED VCPKG_ROOT)`
guard

daschuer commented:
We also need to set the VCPKG_DEFAULT_HOST_TRIPLET
https://github.com/mixxxdj/mixxx/blob/78d2103520e1bec4440b93c120276b652b9ce023/.github/workflows/build.yml#L65

This is our workaround that vcpkg does not fall into cross compile mode. 
It is required to find moc and other host tools.  

This may be also checked by CMAKE_HOST_SYSTEM_PROCESSOR to actually allow cross-compiling.

fwcd commented:
> Which one is actually taken?
Is the one in the CMakeList redundant?

I believe that `CMAKE_OSX_DEPLOYMENT_TARGET` applies to the compilation of Mixxx itself, while the vcpkg setting applies to the dependencies, but I'd have to check too.

fwcd commented:
From my experiments we actually do need to set `CMAKE_OSX_DEPLOYMENT_TARGET`, since `clang` will otherwise be invoked with `x86_64-apple-macosx12.0.0` as triple (or whichever macOS version is used as host) when compiling Mixxx:

![Screen Shot 2022-06-16 at 16 10 35](https://user-images.githubusercontent.com/30873659/174089737-dee4f4d9-1f66-4fdc-9a35-ec7f2cae0a00.png)

With `CMAKE_OSX_DEPLOYMENT_TARGET` explicitly set (like in this PR) `clang` uses the correct triple:

![Screen Shot 2022-06-16 at 16 13 29](https://user-images.githubusercontent.com/30873659/174089881-a988147c-364e-4302-914f-c870b91a4aa8.png)

fwcd committed:
Remove FORCE from CMAKE_OSX_DEPLOYMENT_TARGET

This makes it possible to override the deployment target e.g. via CLI
flags. From my tests, clang still correctly uses the specified
deployment target by default if not overridden.

fwcd commented on CMakeLists.txt:
```diff
@@ -89,10 +89,16 @@ option(QT6 "Build with Qt6" OFF)
+    if(NOT VCPKG_TARGET_TRIPLET)
+      set(VCPKG_TARGET_TRIPLET "x64-osx-min10.15")
+    endif()
   else()
     # Minimum macOS version supported by Qt 5.12
-    set(CMAKE_OSX_DEPLOYMENT_TARGET 10.12 CACHE STRING "Minimum macOS version the build will be able to run on" FORCE)
+    set(CMAKE_OSX_DEPLOYMENT_TARGET 10.12 CACHE STRING "Minimum macOS version the build will be able to run on")
```
I have removed the `FORCE` here since it doesn't seem to be needed and using it prevents overriding it via CLI flags (which e.g. can be useful for building using newer macOS SDKs as e.g. needed for the ARM-native build)

daschuer replied:
We may consider to join that with the above 
```
if(DEFINED ENV{VCPKG_DEFAULT_TRIPLET})
    set(VCPKG_TARGET_TRIPLET "$ENV{VCPKG_DEFAULT_TRIPLET}")
  endif()
```

can you also remove this: 
https://github.com/mixxxdj/mixxx/blob/935ba94f7e498eebd7c3710b2f0df35d989d9a7d/.github/workflows/build.yml#L64

fwcd replied:
> We may consider to join that with the above

I thought about that but figured it would make sense to group it with the other Apple-specific configuration instead. Wdyt?

daschuer replied:
Ok, than let's keep it here.

daschuer commented:
> From my experiments we actually do need to set CMAKE_OSX_DEPLOYMENT_TARGET

I can confirm that. 

We already set the VCPKG_TARGET_TRIPLET a few lines above. 
Can you integrate that?

daschuer commented:
I can also confirm that there is no need to set the  VCPKG_HOST_TRIPLET

See: 
https://github.com/mixxxdj/vcpkg/blob/06db5966107d0f0e733857c04402f60b9ca78df2/scripts/buildsystems/vcpkg.cmake#L543

In our cache we have: 
VCPKG_SETUP_CMAKE_PROGRAM_PATH:BOOL=ON 

This adds the tools directory to the CMAKE_PROGRAM_PATH used for find_program()

Without VCPKG_HOST_TRIPLET set, the target triplet tools are used, that is fine for us.

fwcd committed:
Remove vcpkg triplet flags from macOS CI build

These should no longer be needed given that the correct target triplet
should now be set in the CMakeLists automatically.

daschuer reviewed (approved):
LGTM, Thank you.
